cpp/src/arrow/compute/kernels/aggregate_basic.cc

```diff
@@ -133,6 +134,158 @@ Result<std::unique_ptr<KernelState>> MeanInit(KernelContext* ctx,
   return visitor.Create();
 }
 
+// ----------------------------------------------------------------------
+// Product implementation
+
+using arrow::internal::MultiplyWithOverflow;
+
+template <typename ArrowType>
+struct ProductImpl : public ScalarAggregator {
+  using ThisType = ProductImpl<ArrowType>;
+  using CType = typename ArrowType::c_type;
+  using ProductType = typename FindAccumulatorType<ArrowType>::Type;
+  using OutputType = typename TypeTraits<ProductType>::ScalarType;
+
+  explicit ProductImpl(const ScalarAggregateOptions& options) { this->options = options; }
+
+  Status Consume(KernelContext*, const ExecBatch& batch) override {
+    return ConsumeImpl(batch);
+  }
+
+  template <typename T = ArrowType>
+  enable_if_integer<T, Status> ConsumeImpl(const ExecBatch& batch) {
+    if (batch[0].is_array()) {
+      const auto& data = batch[0].array();
+      this->count += data->length - data->GetNullCount();
+      return VisitArrayDataInline<T>(
+          *data,
+          [&](typename TypeTraits<T>::CType value) {
+            if (ARROW_PREDICT_FALSE(
+                    MultiplyWithOverflow(this->product, value, &this->product))) {
+              return Status::Invalid("Overflow in product");
```

Comment:
Is it possible to provide a sfina `multiply` function for integers and float, so we can eliminate duplicated array visiting code?
The float multiply always returns true, the compiler will optimize away the return value checking.

Reply:
Maybe it's not necessary to check overflow? Call `MultiplyWithOverflow` without checking the return value just to make sure signed integer overflow (UB) will not happen.
`numpy.prod` doesn't check overflow, https://numpy.org/doc/stable/reference/generated/numpy.prod.html.

Reply:
Sounds good to me, thanks for the suggestion. We can add a `hash_product_checked` later if there is need for a checked version.